File: packages/isl-runtime-go/trace.go
// Package islruntime provides runtime verification helpers for ISL (Intent Specification Language)
package islruntime

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strings"
	"time"

	"github.com/google/uuid"
)

// TraceEventType represents the type of trace event
type TraceEventType string

const (
	TraceEventTypeCall        TraceEventType = "call"
	TraceEventTypeReturn      TraceEventType = "return"
	TraceEventTypeStateChange TraceEventType = "state_change"
	TraceEventTypeCheck       TraceEventType = "check"
	TraceEventTypeError       TraceEventType = "error"
)

// TraceEvent represents a single trace event
type TraceEvent struct {
	ID          string                 `json:"id"`
	Type        TraceEventType         `json:"type"`
	Timestamp   int64                  `json:"timestamp"`
	Data        map[string]interface{} `json:"data"`
	Behavior    *string                `json:"behavior,omitempty"`
	Input       map[string]interface{} `json:"input,omitempty"`
	Output      interface{}            `json:"output,omitempty"`
	Error       *ErrorInfo             `json:"error,omitempty"`
	StateBefore *EntityStoreSnapshot   `json:"state_before,omitempty"`
	StateAfter  *EntityStoreSnapshot   `json:"state_after,omitempty"`
}

// ErrorInfo represents error information
type ErrorInfo struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

// EntityStoreSnapshot represents a snapshot of entity state
type EntityStoreSnapshot struct {
	Entities map[string]map[string]interface{} `json:"entities"`
}

// TraceMetadata contains metadata about a trace
type TraceMetadata struct {
	TestName      string  `json:"test_name"`
	Scenario      string  `json:"scenario"`
	Implementation *string `json:"implementation,omitempty"`
	Version       string  `json:"version"`
	Environment   string  `json:"environment"`
	Passed        bool    `json:"passed"`
	FailureIndex  *int    `json:"failure_index,omitempty"`
	Duration      int64   `json:"duration"`
}

// Trace represents a complete trace
type Trace struct {
	ID           string                 `json:"id"`
	Name         string                 `json:"name"`
	Domain       string                 `json:"domain"`
	StartTime    int64                  `json:"start_time"`
	EndTime      int64                  `json:"end_time"`
	Events       []TraceEvent           `json:"events"`
	InitialState map[string]interface{} `json:"initial_state"`
	Snapshots    []interface{}           `json:"snapshots"`
	Metadata     TraceMetadata           `json:"metadata"`
}

// TraceEmitter emits trace events during runtime execution
type TraceEmitter struct {
	traceID      string
	startTime    int64
	events       []TraceEvent
	initialState map[string]interface{}
	domain       string
	behavior     string
	eventCounter int
}

// NewTraceEmitter creates a new trace emitter
func NewTraceEmitter(domain, behavior string) *TraceEmitter {
	return &TraceEmitter{
		traceID:      fmt.Sprintf("trace_%d_%s", time.Now().UnixMilli(), uuid.New().String()),
		startTime:    time.Now().UnixMilli(),
		events:       make([]TraceEvent, 0),
		initialState: make(map[string]interface{}),
		domain:       domain,
		behavior:     behavior,
		eventCounter: 0,
	}
}

// CaptureInitialState captures the initial state
func (te *TraceEmitter) CaptureInitialState(state map[string]interface{}) {
	te.initialState = te.redactPII(state)
}

// EmitCall emits a function call event
func (te *TraceEmitter) EmitCall(functionName string, args map[string]interface{}) {
	redactedArgs := te.redactPII(args)
	te.events = append(te.events, TraceEvent{
		ID:        te.generateEventID(),
		Type:      TraceEventTypeCall,
		Timestamp: time.Now().UnixMilli(),
		Data: map[string]interface{}{
			"kind":     "call",
			"function": functionName,
			"args":     redactedArgs,
		},
		Behavior: &te.behavior,
		Input:    redactedArgs,
	})
}

// EmitReturn emits a function return event
func (te *TraceEmitter) EmitReturn(functionName string, result interface{}, durationMs int64) {
	redactedResult := te.redactValue(result)
	te.events = append(te.events, TraceEvent{
		ID:        te.generateEventID(),
		Type:      TraceEventTypeReturn,
		Timestamp: time.Now().UnixMilli(),
		Data: map[string]interface{}{
			"kind":     "return",
			"function": functionName,
			"result":   redactedResult,
			"duration": durationMs,
		},
		Behavior: &te.behavior,
		Output:   redactedResult,
	})
}

// EmitStateChange emits a state change event
func (te *TraceEmitter) EmitStateChange(path []string, oldValue, newValue interface{}, source string) {
	te.events = append(te.events, TraceEvent{
		ID:        te.generateEventID(),
		Type:      TraceEventTypeStateChange,
		Timestamp: time.Now().UnixMilli(),
		Data: map[string]interface{}{
			"kind":      "state_change",
			"path":      path,
			"oldValue":  te.redactValue(oldValue),
			"newValue":  te.redactValue(newValue),
			"source":    source,
		},
		Behavior: &te.behavior,
	})
}

// EmitCheck emits a check event (precondition, postcondition, invariant)
func (te *TraceEmitter) EmitCheck(expression string, passed bool, category string, expected, actual interface{}, message *string) {
	eventType := TraceEventTypeCheck
	data := map[string]interface{}{
		"kind":      "check",
		"expression": expression,
		"passed":    passed,
		"category":  category,
	}
	if expected != nil {
		data["expected"] = te.redactValue(expected)
	}
	if actual != nil {
		data["actual"] = te.redactValue(actual)
	}
	if message != nil {
		data["message"] = *message
	}

	te.events = append(te.events, TraceEvent{
		ID:        te.generateEventID(),
		Type:      eventType,
		Timestamp: time.Now().UnixMilli(),
		Data:      data,
		Behavior:  &te.behavior,
	})
}

// EmitError emits an error event
func (te *TraceEmitter) EmitError(message string, code *string, stack *string) {
	errorCode := "UNKNOWN"
	if code != nil {
		errorCode = *code
	}
	te.events = append(te.events, TraceEvent{
		ID:        te.generateEventID(),
		Type:      TraceEventTypeError,
		Timestamp: time.Now().UnixMilli(),
		Data: map[string]interface{}{
			"kind":    "error",
			"message": message,
			"code":    errorCode,
			"stack":   te.redactPIIValue(stack),
		},
		Behavior: &te.behavior,
		Error: &ErrorInfo{
			Code:    errorCode,
			Message: message,
		},
	})
}

// Finalize finalizes and returns the trace
func (te *TraceEmitter) Finalize(passed bool) *Trace {
	endTime := time.Now().UnixMilli()
	duration := endTime - te.startTime

	return &Trace{
		ID:           te.traceID,
		Name:         fmt.Sprintf("%s - %s", te.domain, te.behavior),
		Domain:       te.domain,
		StartTime:    te.startTime,
		EndTime:      endTime,
		Events:       te.events,
		InitialState: te.initialState,
		Snapshots:    make([]interface{}, 0),
		Metadata: TraceMetadata{
			TestName:    fmt.Sprintf("%s::%s", te.domain, te.behavior),
			Scenario:    te.behavior,
			Version:     "1.0.0",
			Environment: "runtime",
			Passed:      passed,
			Duration:    duration,
		},
	}
}

// SaveToFile saves the trace to a file
func (te *TraceEmitter) SaveToFile(path string, passed bool) error {
	trace := te.Finalize(passed)
	data, err := json.MarshalIndent(trace, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal trace: %w", err)
	}
	return os.WriteFile(path, data, 0644)
}

func (te *TraceEmitter) generateEventID() string {
	te.eventCounter++
	return fmt.Sprintf("evt_%d_%d", te.eventCounter, time.Now().UnixMilli())
}

func (te *TraceEmitter) redactPII(value map[string]interface{}) map[string]interface{} {
	redacted := make(map[string]interface{})
	for k, v := range value {
		lowerKey := strings.ToLower(k)
		if te.isForbiddenKey(lowerKey) {
			continue
		}
		if strings.Contains(lowerKey, "email") {
			if str, ok := v.(string); ok {
				redacted[k] = te.redactEmail(str)
			} else {
				redacted[k] = te.redactValue(v)
			}
		} else if strings.Contains(lowerKey, "ip") || lowerKey == "ip_address" {
			if str, ok := v.(string); ok {
				redacted[k] = te.redactIP(str)
			} else {
				redacted[k] = te.redactValue(v)
			}
		} else if strings.Contains(lowerKey, "phone") {
			if str, ok := v.(string); ok {
				redacted[k] = te.redactPhone(str)
			} else {
				redacted[k] = te.redactValue(v)
			}
		} else {
			redacted[k] = te.redactValue(v)
		}
	}
	return redacted
}

func (te *TraceEmitter) redactValue(value interface{}) interface{} {
	if str, ok := value.(string); ok {
		if strings.Contains(str, "@") && strings.Contains(str, ".") {
			return te.redactEmail(str)
		}
		if te.isIPAddress(str) {
			return te.redactIP(str)
		}
	}
	return value
}

func (te *TraceEmitter) redactPIIValue(value *string) *string {
	if value == nil {
		return nil
	}
	str := *value
	if strings.Contains(str, "@") && strings.Contains(str, ".") {
		redacted := te.redactEmail(str)
		return &redacted
	}
	if te.isIPAddress(str) {
		redacted := te.redactIP(str)
		return &redacted
	}
	return value
}

func (te *TraceEmitter) redactEmail(email string) string {
	parts := strings.Split(email, "@")
	if len(parts) != 2 {
		return "***@***"
	}
	local := parts[0]
	domain := parts[1]
	redactedLocal := "*"
	if len(local) > 1 {
		redactedLocal = string(local[0]) + strings.Repeat("*", min(len(local)-1, 3))
	}
	return fmt.Sprintf("%s@%s", redactedLocal, domain)
}

func (te *TraceEmitter) redactIP(ip string) string {
	parts := strings.Split(ip, ".")
	if len(parts) == 4 {
		return fmt.Sprintf("%s.%s.xxx.xxx", parts[0], parts[1])
	}
	return "xxx.xxx.xxx.xxx"
}

func (te *TraceEmitter) redactPhone(phone string) string {
	if len(phone) > 4 {
		return strings.Repeat("*", len(phone)-4) + phone[len(phone)-4:]
	}
	return "****"
}

func (te *TraceEmitter) isForbiddenKey(key string) bool {
	forbidden := []string{
		"password", "password_hash", "secret", "api_key", "apikey",
		"access_token", "accesstoken", "refresh_token", "refreshtoken",
		"private_key", "privatekey", "credit_card", "creditcard",
		"ssn", "social_security",
	}
	for _, f := range forbidden {
		if strings.Contains(key, f) {
			return true
		}
	}
	return false
}

func (te *TraceEmitter) isIPAddress(str string) bool {
	ipRegex := regexp.MustCompile(`^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$`)
	return ipRegex.MatchString(str)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: packages/isl-runtime-rs/src/trace.rs
//! Trace event emitter for runtime verification

use crate::types::*;
use chrono::Utc;
use serde_json::json;
use std::collections::HashMap;
use uuid::Uuid;

/// Emits trace events during runtime execution
pub struct TraceEmitter {
    trace_id: String,
    start_time: i64,
    events: Vec<TraceEvent>,
    initial_state: serde_json::Value,
    domain: String,
    behavior: String,
    event_counter: usize,
}

impl TraceEmitter {
    /// Create a new trace emitter
    pub fn new(domain: impl Into<String>, behavior: impl Into<String>) -> Self {
        Self {
            trace_id: format!("trace_{}_{}", Utc::now().timestamp_millis(), Uuid::new_v4()),
            start_time: Utc::now().timestamp_millis(),
            events: Vec::new(),
            initial_state: json!({}),
            domain: domain.into(),
            behavior: behavior.into(),
            event_counter: 0,
        }
    }

    /// Capture initial state
    pub fn capture_initial_state(&mut self, state: serde_json::Value) {
        self.initial_state = self.redact_pii(state);
    }

    /// Emit a function call event
    pub fn emit_call(&mut self, function_name: &str, args: &serde_json::Value) {
        self.events.push(TraceEvent {
            id: self.generate_event_id(),
            event_type: TraceEventType::Call,
            timestamp: Utc::now().timestamp_millis(),
            data: json!({
                "kind": "call",
                "function": function_name,
                "args": self.redact_pii(args.clone()),
            }),
            behavior: Some(self.behavior.clone()),
            input: Some(self.redact_pii(args.clone())),
            output: None,
            error: None,
            state_before: None,
            state_after: None,
        });
    }

    /// Emit a function return event
    pub fn emit_return(&mut self, function_name: &str, result: &serde_json::Value, duration_ms: i64) {
        self.events.push(TraceEvent {
            id: self.generate_event_id(),
            event_type: TraceEventType::Return,
            timestamp: Utc::now().timestamp_millis(),
            data: json!({
                "kind": "return",
                "function": function_name,
                "result": self.redact_value(result.clone()),
                "duration": duration_ms,
            }),
            behavior: Some(self.behavior.clone()),
            input: None,
            output: Some(self.redact_value(result.clone())),
            error: None,
            state_before: None,
            state_after: None,
        });
    }

    /// Emit a state change event
    pub fn emit_state_change(
        &mut self,
        path: &[String],
        old_value: &serde_json::Value,
        new_value: &serde_json::Value,
        source: &str,
    ) {
        self.events.push(TraceEvent {
            id: self.generate_event_id(),
            event_type: TraceEventType::StateChange,
            timestamp: Utc::now().timestamp_millis(),
            data: json!({
                "kind": "state_change",
                "path": path,
                "oldValue": self.redact_value(old_value.clone()),
                "newValue": self.redact_value(new_value.clone()),
                "source": source,
            }),
            behavior: Some(self.behavior.clone()),
            input: None,
            output: None,
            error: None,
            state_before: None,
            state_after: None,
        });
    }

    /// Emit a check event (precondition, postcondition, invariant)
    pub fn emit_check(
        &mut self,
        expression: &str,
        passed: bool,
        category: &str,
        expected: Option<&serde_json::Value>,
        actual: Option<&serde_json::Value>,
        message: Option<&str>,
    ) {
        let event_type = match category {
            "precondition" => TraceEventType::Check,
            "postcondition" => TraceEventType::Check,
            "invariant" => TraceEventType::Check,
            _ => TraceEventType::Check,
        };

        self.events.push(TraceEvent {
            id: self.generate_event_id(),
            event_type,
            timestamp: Utc::now().timestamp_millis(),
            data: json!({
                "kind": "check",
                "expression": expression,
                "passed": passed,
                "expected": expected.map(|v| self.redact_value(v.clone())),
                "actual": actual.map(|v| self.redact_value(v.clone())),
                "message": message,
                "category": category,
            }),
            behavior: Some(self.behavior.clone()),
            input: None,
            output: None,
            error: None,
            state_before: None,
            state_after: None,
        });
    }

    /// Emit an error event
    pub fn emit_error(&mut self, message: &str, code: Option<&str>, stack: Option<&str>) {
        self.events.push(TraceEvent {
            id: self.generate_event_id(),
            event_type: TraceEventType::Error,
            timestamp: Utc::now().timestamp_millis(),
            data: json!({
                "kind": "error",
                "message": message,
                "code": code,
                "stack": stack.map(|s| self.redact_pii_value(s)),
            }),
            behavior: Some(self.behavior.clone()),
            input: None,
            output: None,
            error: Some(ErrorInfo {
                code: code.unwrap_or("UNKNOWN").to_string(),
                message: message.to_string(),
            }),
            state_before: None,
            state_after: None,
        });
    }

    /// Finalize and return the trace
    pub fn finalize(&self, passed: bool) -> Trace {
        let end_time = Utc::now().timestamp_millis();
        let duration = end_time - self.start_time;

        Trace {
            id: self.trace_id.clone(),
            name: format!("{} - {}", self.domain, self.behavior),
            domain: self.domain.clone(),
            start_time: self.start_time,
            end_time,
            events: self.events.clone(),
            initial_state: self.initial_state.clone(),
            snapshots: vec![],
            metadata: TraceMetadata {
                test_name: format!("{}::{}", self.domain, self.behavior),
                scenario: self.behavior.clone(),
                implementation: None,
                version: "1.0.0".to_string(),
                environment: "runtime".to_string(),
                passed,
                failure_index: None,
                duration,
            },
        }
    }

    /// Save trace to file
    pub fn save_to_file(&self, path: &str, passed: bool) -> Result<(), Box<dyn std::error::Error>> {
        let trace = self.finalize(passed);
        let json = serde_json::to_string_pretty(&trace)?;
        std::fs::write(path, json)?;
        Ok(())
    }

    fn generate_event_id(&mut self) -> String {
        self.event_counter += 1;
        format!("evt_{}_{}", self.event_counter, Utc::now().timestamp_millis())
    }

    fn redact_pii(&self, value: serde_json::Value) -> serde_json::Value {
        match value {
            serde_json::Value::Object(map) => {
                let mut redacted = serde_json::Map::new();
                for (key, val) in map {
                    let lower_key = key.to_lowercase();
                    if self.is_forbidden_key(&lower_key) {
                        continue;
                    }
                    if lower_key.contains("email") {
                        redacted.insert(key, json!(self.redact_email(&val.to_string())));
                    } else if lower_key.contains("ip") || lower_key == "ip_address" {
                        redacted.insert(key, json!(self.redact_ip(&val.to_string())));
                    } else if lower_key.contains("phone") {
                        redacted.insert(key, json!(self.redact_phone(&val.to_string())));
                    } else {
                        redacted.insert(key, self.redact_value(val));
                    }
                }
                serde_json::Value::Object(redacted)
            }
            serde_json::Value::Array(arr) => {
                serde_json::Value::Array(arr.into_iter().map(|v| self.redact_value(v)).collect())
            }
            _ => self.redact_value(value),
        }
    }

    fn redact_value(&self, value: serde_json::Value) -> serde_json::Value {
        if let Some(s) = value.as_str() {
            if s.contains('@') && s.contains('.') {
                return json!(self.redact_email(s));
            }
            if s.matches('.').count() == 3 && s.chars().all(|c| c.is_ascii_digit() || c == '.') {
                return json!(self.redact_ip(s));
            }
        }
        value
    }

    fn redact_pii_value(&self, value: &str) -> String {
        if value.contains('@') && value.contains('.') {
            return self.redact_email(value);
        }
        if value.matches('.').count() == 3 && value.chars().all(|c| c.is_ascii_digit() || c == '.') {
            return self.redact_ip(value);
        }
        value.to_string()
    }

    fn redact_email(&self, email: &str) -> String {
        if let Some(at_pos) = email.find('@') {
            let (local, domain) = email.split_at(at_pos);
            let redacted_local = if local.len() > 1 {
                format!("{}{}", &local[..1], "*".repeat((local.len() - 1).min(3)))
            } else {
                "*".to_string()
            };
            format!("{}@{}", redacted_local, domain.trim_start_matches('@'))
        } else {
            "***@***".to_string()
        }
    }

    fn redact_ip(&self, ip: &str) -> String {
        let parts: Vec<&str> = ip.split('.').collect();
        if parts.len() == 4 {
            format!("{}.{}.xxx.xxx", parts[0], parts[1])
        } else {
            "xxx.xxx.xxx.xxx".to_string()
        }
    }

    fn redact_phone(&self, phone: &str) -> String {
        if phone.len() > 4 {
            format!("{}{}", "*".repeat(phone.len() - 4), &phone[phone.len() - 4..])
        } else {
            "****".to_string()
        }
    }

    fn is_forbidden_key(&self, key: &str) -> bool {
        let forbidden = [
            "password", "password_hash", "secret", "api_key", "apikey",
            "access_token", "accesstoken", "refresh_token", "refreshtoken",
            "private_key", "privatekey", "credit_card", "creditcard",
            "ssn", "social_security",
        ];
        forbidden.iter().any(|f| key.contains(f))
    }
}
